streaming/cleanup.go

```diff
@@ -0,0 +1,315 @@
+		}
+		//cleanup
+		for _, resources := range datastreamResourcesList {
+			var datastreamResources internal.DatastreamResources
+			err := json.Unmarshal([]byte(resources.ResourceMetadata), &datastreamResources)
+			if err != nil {
+				logger.Log.Debug("Unable to read Datastream metadata for deletion\n")
+			} else {
+				cleanupDatastream(ctx, datastreamResources, project)
+			}
+		}
+	}
+	if jobCleanupOptions.Pubsub {
+		//fetch pubsub resources
+		pubsubResourcesList, err := FetchResources(ctx, migrationJobId, constants.PUBSUB_RESOURCE, dataShardIds, project, instance)
+		if err != nil {
+			logger.Log.Debug(fmt.Sprintf("Unable to fetch pubsub resources for jobId: %s: %v\n", migrationJobId, err))
+		}
+		//cleanup
+		for _, resources := range pubsubResourcesList {
+			var pubsubResources internal.PubsubResources
+			err := json.Unmarshal([]byte(resources.ResourceMetadata), &pubsubResources)
+			if err != nil {
+				logger.Log.Debug("Unable to read Pubsub metadata for deletion\n")
+			} else {
+				cleanupPubsubResources(ctx, pubsubResources, project)
+			}
+		}
+	}
+	if jobCleanupOptions.Monitoring {
+		//fetch monitoring resources
+		shardMonitoringResourcesList, err := FetchResources(ctx, migrationJobId, constants.MONITORING_RESOURCE, dataShardIds, project, instance)
+		if err != nil {
+			logger.Log.Debug(fmt.Sprintf("Unable to fetch shard monitoring resources for jobId: %s: %v\n", migrationJobId, err))
+		}
+		jobMonitoringResourcesList, err := FetchResources(ctx, migrationJobId, constants.AGG_MONITORING_RESOURCE, dataShardIds, project, instance)
+		if err != nil {
+			logger.Log.Debug(fmt.Sprintf("Unable to fetch aggregate monitoring resources for jobId: %s: %v\n", migrationJobId, err))
+		}
+		monitoringResourcesList := append(shardMonitoringResourcesList, jobMonitoringResourcesList...)
+		//cleanup
+		for _, resources := range monitoringResourcesList {
+			var monitoringResources internal.MonitoringResources
+			err := json.Unmarshal([]byte(resources.ResourceMetadata), &monitoringResources)
+			if err != nil {
+				logger.Log.Debug("Unable to read monitoring metadata for deletion\n")
+			} else {
+				cleanupMonitoringDashboard(ctx, monitoringResources, project)
+			}
+		}
+	}
+}
+
+func FetchResources(ctx context.Context, migrationJobId string, resourceType string, dataShardIds []string, project string, instance string) ([]JobResources, error) {
+	dbURI := fmt.Sprintf("projects/%s/instances/%s/databases/%s", project, instance, constants.METADATA_DB)
+	client, err := utils.GetClient(ctx, dbURI)
+	if err != nil {
+		err = fmt.Errorf("can't create client for db %s: %v", dbURI, err)
+		return nil, err
+	}
+	defer client.Close()
+	txn := client.ReadOnlyTransaction()
+	defer txn.Close()
+
+	//fetch all resources
+	var resourceQuery spanner.Statement
+	if dataShardIds != nil {
+		//query the provided data shards only
+		resourceQuery = spanner.Statement{
+			SQL: `SELECT 
```

Comment:
nit: both the queries can probably be merged adding the data shards where required. This will avoid changes in 2 places going forward.

Reply:
Didn't get you, do you mean extracting out the common part of the query in a separate string? 

I have currently kept it this way because extracting out the common part would decrease readability IMO. Open to suggestions!